driver/controller.go

```diff
@@ -251,26 +279,72 @@ func (d *Driver) ControllerUnpublishVolume(ctx context.Context, req *csi.Control
 	})
 	ll.Info("controller unpublish volume called")
 
+	// check if volume exist before trying to detach it
+	volume, err := d.cloudscaleClient.Volumes.Get(ctx, req.VolumeId)
+	if err != nil {
+		errorResponse, ok := err.(*cloudscale.ErrorResponse)
+		if ok && errorResponse.StatusCode == http.StatusNotFound {
+			d.log.WithFields(logrus.Fields{
+				"volume_id": req.VolumeId,
+				"node_id":   req.NodeId,
+				"method":    "controller_unpublish_volume",
+			}).Infof("volume %q not found; assuming it's already detached", req.VolumeId)
+			// assume it's detached
+			return &csi.ControllerUnpublishVolumeResponse{}, nil
+		}
+		d.log.WithFields(logrus.Fields{
+			"volume_id": req.VolumeId,
+			"node_id":   req.NodeId,
+			"method":    "controller_unpublish_volume",
+		}).Warnf("unable to query status of volume %q", req.VolumeId)
+		return nil, err
+	}
+
+	// check if volume is already detached
+	if volume.ServerUUIDs != nil && len(*volume.ServerUUIDs) == 0 {
+		d.log.WithFields(logrus.Fields{
+			"volume_id": req.VolumeId,
+			"volume":    volume,
+			"node_id":   req.NodeId,
+			"method":    "controller_unpublish_volume",
+		}).Infof("volume %q is not attached to any servers", req.VolumeId)
+		return &csi.ControllerUnpublishVolumeResponse{}, nil
+	}
+
 	detachRequest := &cloudscale.Volume{
 		ServerUUIDs: &[]string{},
 	}
-	err := d.cloudscaleClient.Volumes.Update(ctx, req.VolumeId, detachRequest)
+	err = d.cloudscaleClient.Volumes.Update(ctx, req.VolumeId, detachRequest)
 	if err != nil {
-		/*
-			if resp != nil && resp.StatusCode == http.StatusUnprocessableEntity {
-				if strings.Contains(err.Error(), "Droplet already has a pending event") {
-					ll.WithFields(logrus.Fields{
-						"error": err,
-						"resp":  resp,
-					}).Warn("droplet is not able to detach the volume")
-
-					// sending an abort makes sure the csi-attacher retries with the next backoff tick
-					return nil, status.Errorf(codes.Aborted, "volume %q couldn't be detached. droplet %d is in process of another action",
-						req.VolumeId, dropletID)
-				}
-			}
-		*/
```

Comment:
I guess here I have the same question as above.

Reply:
In this case, semantics would change if we were using `reraiseNotFound`. 

If the volume no longer exists, it must have already been detached previously, maybe through the cloudscale.ch UI or API calls by an administrator.  We must not return an error in this case to ensure that the operations are idemtpotent. 

I'm not a big fan of the verbosity the changed code introduces into the codebase, but it makes the code easy to read and the return values are explicitly visible, which is not the case with `reraiseNotFound`.

Reply:
In this case, semantics would change if we were using `reraiseNotFound`.

If the volume no longer exists, it must have already been detached previously, maybe through the cloudscale.ch UI or API calls by an administrator. We must not return an error in this case to ensure that the operations are idemtpotent.

I'm not a big fan of the verbosity the changed code introduces into the codebase, but it makes the code easy to read and the return values are explicitly visible, which is not the case with `reraiseNotFound`.